File: competition/07-tree/tree.js
function leaf(v) {
  return { tag: "leaf", v };
}

function node(l, r) {
  return { tag: "node", l, r };
}

function build(depth, start) {
  if (depth === 0) return leaf(start);
  return node(build(depth - 1, start), build(depth - 1, start + 1));
}

function treeSum(t) {
  if (t.tag === "leaf") return t.v;
  return treeSum(t.l) + treeSum(t.r);
}

console.log(treeSum(build(21, 1)));

File: competition/07-tree/tree.py
import sys

sys.setrecursionlimit(1 << 20)


class Leaf:
    __slots__ = ("v",)

    def __init__(self, v):
        self.v = v


class Node:
    __slots__ = ("l", "r")

    def __init__(self, l, r):
        self.l = l
        self.r = r


def build(depth, start):
    if depth == 0:
        return Leaf(start)
    return Node(build(depth - 1, start), build(depth - 1, start + 1))


def tree_sum(t):
    if isinstance(t, Leaf):
        return t.v
    return tree_sum(t.l) + tree_sum(t.r)


print(tree_sum(build(21, 1)))
